src/github.com/matrix-org/dendrite/mediaapi/writers/fileutils.go

```diff
@@ -0,0 +1,287 @@
+// Returns the file handle as it needs to be closed when writing is complete
+func createFileWriter(directory types.Path, filename types.Filename) (*bufio.Writer, *os.File, error) {
+	filePath := path.Join(string(directory), string(filename))
+	file, err := os.Create(filePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create file: %v", err)
+	}
+
+	return bufio.NewWriter(file), file, nil
+}
+
+func createTempFileWriter(absBasePath types.Path) (*bufio.Writer, *os.File, types.Path, error) {
+	tmpDir, err := createTempDir(absBasePath)
+	if err != nil {
+		return nil, nil, "", fmt.Errorf("Failed to create temp dir: %q", err)
+	}
+	writer, tmpFile, err := createFileWriter(tmpDir, "content")
+	if err != nil {
+		return nil, nil, "", fmt.Errorf("Failed to create file writer: %q", err)
+	}
+	return writer, tmpFile, tmpDir, nil
+}
+
+var errFileIsTooLarge = fmt.Errorf("file is too large")
+var errRead = fmt.Errorf("failed to read response from remote server")
+var errResponse = fmt.Errorf("failed to write file data to response body")
+var errHash = fmt.Errorf("failed to hash file data")
+var errWrite = fmt.Errorf("failed to write file to disk")
+
+// writeToResponse takes bytesToWrite bytes from buffer and writes them to respWriter
+// Returns bytes written and an error. In case of error, or if there is no respWriter,
+// the number of bytes written will be 0.
+func writeToResponse(respWriter http.ResponseWriter, buffer []byte, bytesToWrite int) (int64, error) {
+	if respWriter != nil {
+		bytesWritten, respErr := respWriter.Write(buffer[:bytesToWrite])
+		if bytesWritten != bytesToWrite || (respErr != nil && respErr != io.EOF) {
+			return 0, errResponse
+		}
+		return int64(bytesWritten), nil
+	}
+	return 0, nil
+}
+
+// writeToDiskAndHasher takes bytesToWrite bytes from buffer and writes them to tmpFileWriter and hasher.
+// Returns bytes written and an error. In case of error, including if writing would exceed maxFileSizeBytes,
+// the number of bytes written will be 0.
+func writeToDiskAndHasher(tmpFileWriter *bufio.Writer, hasher hash.Hash, bytesWritten int64, maxFileSizeBytes types.ContentLength, buffer []byte, bytesToWrite int) (int64, error) {
+	// if larger than maxFileSizeBytes then stop writing to disk and discard cached file
+	if bytesWritten+int64(bytesToWrite) > int64(maxFileSizeBytes) {
+		return 0, errFileIsTooLarge
+	}
+	// write to hasher and to disk
+	bytesTemp, writeErr := tmpFileWriter.Write(buffer[:bytesToWrite])
+	bytesHashed, hashErr := hasher.Write(buffer[:bytesToWrite])
+	if writeErr != nil && writeErr != io.EOF || bytesTemp != bytesToWrite || bytesTemp != bytesHashed {
+		return 0, errWrite
+	} else if hashErr != nil && hashErr != io.EOF {
+		return 0, errHash
+	}
+	return int64(bytesTemp), nil
+}
+
+// readAndHashAndWriteWithLimit works like io.Copy except it copies from the reqReader to the
+// optionally-supplied respWriter and a temporary file named 'content' using a bufio.Writer.
```

Comment:
The first line of function-level docs usually gives a nice handy one-line summary of the function, before going into more details with function args where necessary. This is done to provide a high-level overview (moreso than the function name). The docs for this function seem all over the place. This may be because this function is trying to do too much. From what I can tell:
 - It works out a temporary file where it needs to write to.
 - It writes `reqReader` to disk and simultaneously hashes it.
 - It limits the amount written to `maxFileSizeBytes`.
 - It optionally proxies the data that was written to disk to a writer.

It sounds like this is something you'd want to do on `/upload` and when fetching data from federated homeservers, so factoring it out makes sense. I think this is just a docs/API shape thing. It would be great if you could re-write these comments to be a bit simpler (eg by using those 4 bullet points as a high-level overview).

The function name is also not great. I don't need to know in the function name that I can apply a limit, I can tell that from the function arg `maxFileSizeBytes`. I don't need to know in the function name that you're going to read the reader I provide to you, I can reasonably assume that (what else are you going to do with it? 😛 ). The main point of this function is to write things to disk and to work out a hash in the process, so perhaps `writeTempFile(reader, limit, basePath, writer)` would be clearer. We already document that there is going to be a base64 hash returned (it's a return arg), so a one-liner to say "oh by the way, this hash is the hash of the bytes written" would be plenty.